File: 46-permutations/46-permutations.py
class Solution:
    '''
    p([1,2], [])
        p([1,2], [1])
            p([1,2], [1,2])
        p([1,2], [2])
            p([1,2], [2,1])
    '''
    def __init__(self):
        self.ans = []
    def helper(self, nums, cur):
        if len(cur) == len(nums):
            self.ans.append(cur)
            return
        for num in nums:
            if num not in cur:
                # cur.append(num) # Choose
                self.helper(nums, cur + [num]) # Explore
                # cur.pop() # Unchoose
        
    def permute(self, nums, cur=[]):       
        self.helper(nums, [])
        return self.ans

File: 46-permutations/46-permutations.java
class Solution {
    
    public static void permutations(int[] nums, int size, List<Integer> yet, List<List<Integer>> ans){
        if (yet.size() == nums.length){ 
            ans.add(new ArrayList<Integer>(yet)); 
        }
        
        for (int num: nums){
            if (!yet.contains(num)){
                yet.add(num);
                permutations(nums, size, yet, ans);
                yet.remove(yet.size()-1);
            }
        }
    }
    
    
    public List<List<Integer>> permute(int[] nums) {
        
        List<List<Integer>> ans = new ArrayList<List<Integer>>();
        
        permutations(nums, 0, new ArrayList<Integer>(), ans);
        
        return ans;
        
        
    }
}